test/torchaudio_unittest/rnnt/autograd_impl.py

```diff
@@ -0,0 +1,80 @@
+from typing import Callable, Tuple
+import torch
+from torch import Tensor
+from torch.autograd import gradcheck
+from torchaudio_unittest.common_utils import (
+    TestBaseMixin,
+)
+from torchaudio.prototype.rnnt_loss import RNNTLoss
+from parameterized import parameterized
+from .utils import (
+    numpy_to_torch,
+    # get_B1_T10_U3_D4_data,
+    get_numpy_data_B2_T4_U3_D3,
+    get_numpy_data_B1_T2_U3_D5
+)
+from .numpy_transducer import NumpyTransducerLoss
+
+
+class Autograd(TestBaseMixin):
+    @staticmethod
+    def get_data(data_func, device):
+        data_np = data_func()
+        if type(data_np) == tuple:
+            print("reference gradient")
+            print(data_np[-1])
+            data_np = data_np[0]
+        data = numpy_to_torch(
+            data=data_np, device=device, requires_grad=True
+        )
+        return data
+
+    def assert_grad(
+            self,
+            loss: Callable[..., Tensor],
+            inputs: Tuple[torch.Tensor],
+            *,
+            enable_all_grad: bool = True,
+    ):
+        # inputs_ = []
+        # for i in inputs:
+        #     if torch.is_tensor(i):
+        #         i = i.to(dtype=self.dtype, device=self.device)
+        #         if enable_all_grad:
+        #             i.requires_grad = True
+        #     inputs_.append(i)
+        assert gradcheck(loss, inputs, eps=1e-03, atol=1e-02, rtol=1e-02, nondet_tol=0.)
+
+    @parameterized.expand([
+        # (get_B1_T10_U3_D4_data, ),
+        (get_numpy_data_B2_T4_U3_D3, ),
+        (get_numpy_data_B1_T2_U3_D5, ),
+    ])
+    def test_RNNTLoss_gradcheck(self, data_func):
+        data = self.get_data(data_func, self.device)
+        inputs = (
+            data["logits"].to(self.dtype),
+            data["targets"],
+            data["logit_lengths"],
+            data["target_lengths"],
+        )
+        loss = RNNTLoss(blank=data["blank"], reuse_logits_for_grads=False)
```

Comment:
I can see that tests are also passing on CPU with `fused_log_softmax=False` (because it sets `reuse_logits_for_grads` to `False`) [here](https://github.com/vincentqb/audio/blob/rnntautograd/torchaudio/prototype/rnnt_loss.py#L38):

```python
if not fused_log_softmax:
    logits = torch.nn.functional.log_softmax(logits, dim=-1)
    reuse_logits_for_grads = (
        False  # softmax needs the original logits value
    )
```

What would be the recommendation re which argument should be set to `False` by default?

FYI: `warp-transducer`'s default behavior is to apply `log_softmax` on CPU [here](https://github.com/HawkAaron/warp-transducer/blob/master/pytorch_binding/warprnnt_pytorch/__init__.py#L67):

```python
if not acts.is_cuda:
    acts = torch.nn.functional.log_softmax(acts, -1)
```

P.S. This might be a nit...

Reply:
re warp-transducer: `log_softmax` is part of the CUDA kernel, but not part of the CPU kernel in C++. hence, on CPU, they apply `log_softmax` within python. is that what you meant?

re default: that's a good catch. i'd follow the CTC convention, which doesn't apply softmax, [here](https://pytorch.org/docs/stable/generated/torch.nn.CTCLoss.html#torch.nn.CTCLoss). thoughts?

Reply:
> that's a good catch for fused_log_softmax. i'd follow the CTC convention, which doesn't apply softmax, here. thoughts?

up to @carolineechen , @mthrok and @cpuhrsch 

> for warp-transducer: log_softmax is part of the CUDA kernel, but not part of the CPU kernel in C++. hence, on CPU, they apply log_softmax within python. is that what you meant?

yes

Reply:
quick clarification: CTC doesn't apply `log_softmax`, and we should decide whether to do the same when `fused_log_softmax=False` since it currently does.

Reply:
What I expect from a loss function is that I can perform `.backward()` on the output values. It should not equire any processing after the loss function. Otherwise it should not be called loss. 

And since applying softmax and NLL loss has numerical issue, I think cross entropy loss should be applied here. https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html

Reply:
Can we add a test to ensure that the result of CPU a d CUDA are reasonably close?

Reply:
> What I expect from a loss function is that I can perform `.backward()` on the output values. It should not equire any processing after the loss function. Otherwise it should not be called loss.
> 
> Assuming that NLL is applied after softmax, since applying softmax and NLL loss has numerical issue, I think cross entropy loss should be applied here. https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html

Given that [warp-transducer](https://github.com/HawkAaron/warp-transducer/blob/master/pytorch_binding/warprnnt_pytorch/__init__.py#L67) and [warp-rnnt](https://github.com/1ytic/warp-rnnt/blob/master/pytorch_binding/warp_rnnt/__init__.py#L38) always apply `log_softmax`, I agree that RNNTL is akin to CrossEntropy here, and so we should always apply `log_softmax`.

Reply:
Added providing assertion error instead of silently changing the parameter of `reuse_logits_for_grads`

> Can we add a test to ensure that the result of CPU a d CUDA are reasonably close?

Added to follow-ups.

> What I expect from a loss function is that I can perform `.backward()` on the output values. It should not equire any processing after the loss function. Otherwise it should not be called loss.
> 
> Assuming that NLL is applied after softmax, since applying softmax and NLL loss has numerical issue, I think cross entropy loss should be applied here. https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html

I summarize the conventions for the loss as:
- CTC does not apply `log_softmax`.
- NLL offers a fused version of NLL and `log_softmax`.

[warp-transducer](https://github.com/HawkAaron/warp-transducer/blob/master/pytorch_binding/warprnnt_pytorch/__init__.py#L67) and [warp-rnnt](https://github.com/1ytic/warp-rnnt/blob/master/pytorch_binding/warp_rnnt/__init__.py#L38) do the latter. However, fused is not compatible with `reuse_logits_for_grads`.

Based on the discussion above, we'll keep the path of `fused_log_softmax=True` and `reuse_logits_for_grads=False` but remove both parameters from the signature.

Once we have validated the benefits of exposing `reuse_logits_for_grads=True` (with `fused_log_softmax=False`), we can revisit this discussion. My preliminary suggestion will be to add a new python function (say `rnn_loss_without_log_softmax_`) to indicate a version of RNNTL without log_softmax done inplace.

Reply:
@vincentqb @carolineechen 

What is the benefit and the necessity for `fused_log_softmax=False`?
Our decision for that API should be based on the value and usability of the option, not on the convention of the existing ones. We first need to check the necessity and the usecases. We can first provide API without it and added it later if there is a user demand. Adding parameter is easier than removing, and if we are adding that APIs we do need to consider the maintenance cost.

Reply:
> What is the benefit and the necessity for `fused_log_softmax=False`?

`reuse_logits_for_grads=True` requires `fused_log_softmax=False`.

> Our decision for that API should be based on the value and usability of the option, not on the convention of the existing ones. We first need to check the necessity and the usecases. We can first provide API without it and added it later if there is a user demand. Adding parameter is easier than removing, and if we are adding that APIs we do need to consider the maintenance cost.

Indeed, we plan to remove those two parameters. If we find evidence later of the value of `reuse_logits_for_grads` (e.g. memory saved), we'll revisit this question.

Reply:
Since we are in alignment here, I'll resolve this discussion. Updated [internal](https://docs.google.com/document/d/1pCo_aKGY2y_qx0lNaU03IugoARPg_di44CoFseVwWOM/edit).